storage/bucket.go

```diff
@@ -0,0 +1,347 @@
+	defer b.mu.RUnlock()
+	return b.objects[objName]
+}
+
+// Check of an object name's directory is the specified prefix.
+func matchPrefix(path, prefix string) bool {
+	if prefix == "" {
+		return true
+	}
+	i := strings.LastIndexByte(path, '/')
+	return path[:i+1] == prefix
+}
+
+func (b *Bucket) Objects(prefix string) map[string]*storage.Object {
+	objs := make(map[string]*storage.Object)
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	for objName, obj := range b.objects {
+		if matchPrefix(objName, prefix) {
+			objs[objName] = obj
+		}
+	}
+	return objs
+}
+
+func (b *Bucket) Prefixes() []string {
+	prefixmap := make(map[string]bool)
+	prefixlist := make([]string, 0)
+
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	for path := range b.objects {
+		for strings.HasPrefix(path, b.prefix) {
+			i := strings.LastIndexByte(path, '/')
+			if i < 0 {
+				break
+			}
+			path = path[:i+1]
+			if prefixmap[path] {
+				break
+			}
+			prefixmap[path] = true
+			prefixlist = append(prefixlist, path)
+		}
+	}
+	return prefixlist
+}
+
+func (b *Bucket) Len() int {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return len(b.objects)
+}
+
+func (b *Bucket) addObject(obj *storage.Object) {
+	if obj.Bucket != b.name {
+		panic(fmt.Errorf("adding gs://%s/%s to bucket %s", obj.Bucket, obj.Name, b.name))
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.objects[obj.Name] = obj
+}
+
+func (b *Bucket) addObjects(objs *storage.Objects) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	for _, obj := range objs.Items {
+		if obj.Bucket != b.name {
+			panic(fmt.Errorf("adding gs://%s/%s to bucket %s", obj.Bucket, obj.Name, b.name))
+		}
+		b.objects[obj.Name] = obj
+	}
+}
+
+func (b *Bucket) delObject(objName string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	delete(b.objects, objName)
+}
+
+func (b *Bucket) mkURL(obj interface{}) *url.URL {
+	switch v := obj.(type) {
+	case string:
+		u := b.URL()
+		u.Path = v
+		return u
+	case *storage.Object:
+		u := b.URL()
+		u.Path = v.Name
+		if v.Bucket != "" {
+			u.Host = v.Bucket
+		}
+		return u
+	case *url.URL:
+		return v
+	case nil:
+		return b.URL()
+	default:
+		panic("unknown type")
+	}
+}
+
+func (b *Bucket) apiErr(op string, obj interface{}, e error) error {
+	if _, ok := e.(*googleapi.Error); ok {
+		return &Error{Op: op, URL: b.mkURL(obj).String(), Err: e}
+	}
+	return e
+}
+
+func (b *Bucket) Fetch(ctx context.Context) error {
+	req := b.service.Objects.List(b.name)
+	if b.prefix != "" {
+		req.Prefix(b.prefix)
+	}
+
+	add := func(objs *storage.Objects) error {
+		b.addObjects(objs)
+		plog.Infof("Found %d objects under %s", b.Len(), b.URL())
+		return nil
+	}
+
+	startRequest()
+	defer stopRequest()
+	plog.Noticef("Fetching %s", b.URL())
+
+	return b.apiErr("storage.objects.list", nil, req.Pages(nil, add))
+}
+
+func (b *Bucket) Upload(ctx context.Context, obj *storage.Object, media io.ReadSeeker) error {
+	// Calculate the checksum to enable upload integrity checking.
+	if obj.Crc32c == "" {
+		obj = dupObj(obj) // avoid editing the original
```

Comment:
exactly, that's what I want. The hash should be set in order for the server to reject corrupted uploads.